### PR TITLE
test: Enable the counter_basic_api test on mec172x board

### PR DESCRIPTION
**Pull request by hunterhu3000**
1. enable timer5 on dts
2. Replace all BTMR_Type with struct btmr_regs, to make it compatiable
   on both mec 15 and 17
3. As MEC15 and MEC17 have different ways to access the register, use
   "#if defined" to access counter_cfg->girq_id on different board.
4. Fix a bug, when no device is found and the counter test is not
   acutaly run, it will still return PASS.

Test pass on mec172xevb_assy6906

Signed-off-by: Hu Zhenyu <zhenyu.hu@intel.com>

### CODE REVIEW

**Review by MaureenHelm — DISMISSED** [review]

**Comment by hunterhu3000** [comment]
@albertofloyd This PR is related to https://github.com/zephyrproject-rtos/zephyr/pull/45704. After it is merged, I'll fix the issue of https://github.com/zephyrproject-rtos/zephyr/issues/45647. Please review and comment. Thanks

File: dts/arm/microchip/mec172xnsz.dtsi [hunk 1]
```diff
@@ -421,7 +421,7 @@
 			label = "TIMER_5";
 			max-value = <0xFFFFFFFF>;
 			prescaler = <0>;
-			status = "disabled";
+			status = "okay";
 		};
```
Comment (jvasanth1) [inline dts/arm/microchip/mec172xnsz.dtsi:425]: status okay should be in the board level file?

File: drivers/counter/counter_mchp_xec.c [hunk 2]
```diff
@@ -46,7 +46,7 @@ struct counter_xec_data {
 };
 
 #define COUNTER_XEC_REG_BASE(_dev)			\
-	((BTMR_Type *)					\
+	((struct btmr_regs *)					\
 	 ((const struct counter_xec_config * const)	\
```
Comment (jvasanth1) [inline drivers/counter/counter_mchp_xec.c:50]: need to confirm this change works for MEC15xx as well.

File: soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h [hunk 3]
```diff
@@ -100,6 +103,18 @@
...
+/* 8 <= n <= 26 */
+#define MCHP_GIRQ_SRC_ADDR(n) \
+	((MCHP_ECIA_ADDR + 0x00u) + (((uint32_t)(n) - 8u) * 0x14u))
+
+#define MCHP_GIRQ_ENSET_ADDR(n) \
+	((MCHP_ECIA_ADDR + 0x04u) + (((uint32_t)(n) - 8u) * 0x14u))
+
+/* 8 <= n <= 26 corresponding to GIRQ08, GIRQ09, ..., GIRQ26 */
+#define MCHP_GIRQ_SRC(n) REG32(MCHP_GIRQ_SRC_ADDR(n))
+
```
Comment (jvasanth1) [inline soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h:115]: for MEC17xx we are not using macros to access GIRQ registers. We have functions in interrupt driver for this purpose.

File: drivers/counter/counter_mchp_xec.c (hunk 2 above)
Comment (hunterhu3000) [reply drivers/counter/counter_mchp_xec.c:50]: The code works well on MEC15
BTMR_Type is defined like this in MEC15
```
typedef struct btmr_regs
{
        __IOM uint32_t CNT;     /*!< (@ 0x00000000) BTMR Count            */
        __IOM uint32_t PRLD;    /*!< (@ 0x00000004) BTMR Preload          */
        __IOM uint8_t  STS;     /*!< (@ 0x00000008) BTMR Status           */
        uint8_t RSVDC[3];
        __IOM uint8_t IEN;      /*!< (@ 0x0000000c) BTMR Interrupt Enable */
        uint8_t RSVDD[3];
        __IOM uint32_t CTRL;    /*!< (@ 0x00000010) BTMR Control          */
} BTMR_Type;
```
I tried to use typedef in soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h for MEC 17 soc/arm/microchip_mec/common/reg/mec_timers.h, but the check function says new typedef was not allowed. So I modify it in the driver code.

File: dts/arm/microchip/mec172xnsz.dtsi (hunk 1 above)
Comment (hunterhu3000) [reply dts/arm/microchip/mec172xnsz.dtsi:425]: Agree. I'll move the code to board level dts file.

File: soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h (hunk 3 above)
Comment (hunterhu3000) [reply soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h:115]: Do you mean this?
MCHP_GIRQ_SRC ==> mchp_xec_ecia_girq_src_clr
MCHP_GIRQ_ENSET ==>mchp_xec_ecia_girq_src_en

Do you have an official version of counter_mchp_xec.c which can be used for both MEC15 and MEC17?
If so, could you please upload that file?

**Comment by hunterhu3000** [comment]
@jvasanth1 I have made a new patch, thanks to review.

**Comment by MaureenHelm** [comment]
It's been a long time since CI ran on this PR. Please rebase to rerun CI

**Comment by hunterhu3000** [comment]
> It's been a long time since CI ran on this PR. Please rebase to rerun CI

OK, rebase is done

File: drivers/counter/counter_mchp_xec.c [hunk 4]
```diff
@@ -243,14 +243,19 @@ static int counter_xec_set_top_value(const struct device *dev,
...
-	BTMR_Type *counter = COUNTER_XEC_REG_BASE(dev);
+	struct btmr_regs *counter = COUNTER_XEC_REG_BASE(dev);
 	const struct counter_xec_config *counter_cfg = COUNTER_XEC_CONFIG(dev);
 	struct counter_xec_data *data = COUNTER_XEC_DATA(dev);
 	counter_alarm_callback_t alarm_cb;
 	void *user_data;
 
 	counter->STS = MCHP_BTMR_STS_ACTIVE;
+
+#if defined(CONFIG_BOARD_MEC172XEVB_ASSY6906)
```
Comment (galak) [inline drivers/counter/counter_mchp_xec.c:254]: Having board specific ifdef in driver code should be avoided.

File: drivers/counter/counter_mchp_xec.c [hunk 5]
```diff
@@ -294,7 +299,11 @@ static int counter_xec_init(const struct device *dev)
 	counter->PRLD = counter_cfg->info.max_top_value;
 	counter->CNT = counter_cfg->info.max_top_value;
 
+#if defined(CONFIG_BOARD_MEC172XEVB_ASSY6906)
```
Comment (galak) [inline drivers/counter/counter_mchp_xec.c:302]: Figure out how to do this so its not board specific

**Review by galak — DISMISSED** [review]
replace board specific code in drivers

**Review by albertofloyd — DISMISSED** [review]

File: drivers/counter/counter_mchp_xec.c (hunk 4 above)
Comment (albertofloyd) [reply drivers/counter/counter_mchp_xec.c:254]: @galak MEC152 uses a HAL and MEC172x does not the header definitions are very different.
Any suggestions @jvasanth1

**Comment by hunterhu3000** [comment]
Rebased to the latest and fix the problem in https://github.com/zephyrproject-rtos/zephyr/issues/45647

**Review by albertofloyd — DISMISSED** [review]

File: drivers/counter/counter_mchp_xec.c (hunk 4 above)
Comment (galak) [reply drivers/counter/counter_mchp_xec.c:254]: Is this really board specific or SoC specific?

File: drivers/counter/counter_mchp_xec.c (hunk 5 above)
Comment (galak) [reply drivers/counter/counter_mchp_xec.c:302]: It feels like this is a SoC difference and not a board specific one.  If that is really the case, than I suggest just changing the ifdef to be based on SoC.

Comment (hunterhu3000) [reply drivers/counter/counter_mchp_xec.c]: @galak Do you mean replace CONFIG_BOARD_MEC172XEVB_ASSY6906 with CONFIG_SOC_MEC172X_NSZ?
If yes, I have pushed a new patch and rebased. Thanks to review.

File: soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz [hunk 6]
```diff
@@ -23,4 +23,8 @@ config EEPROM_XEC
 	default y
 	depends on EEPROM
 
+config COUNTER_XEC
+	default y
+	depends on COUNTER
+
```
Comment (MaureenHelm) [inline soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz:29]: This can be removed following 0f1ed0f4a05

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please move the fix in `tests/drivers/counter/counter_basic_api/src/test_counter.c` to a separate commit

File: soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz (hunk 6 above)
Comment (hunterhu3000) [reply soc/arm/microchip_mec/mec172x/Kconfig.defconfig.mec172xnsz]: Yes, it can be removed.
The patch is updated. Thanks to review

**Comment by MaureenHelm** [comment]
> Please move the fix in `tests/drivers/counter/counter_basic_api/src/test_counter.c` to a separate commit

This still needs to be done

**Comment by hunterhu3000** [comment]
> > Please move the fix in `tests/drivers/counter/counter_basic_api/src/test_counter.c` to a separate commit
> 
> This still needs to be done

Done.
The CI fail looks irrelevant to the patches. The patches works well on mec172xevb. Thanks to check.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by hunterhu3000** [comment]
@albertofloyd could you please review and +1? Thanks

**Comment by MaureenHelm** [comment]
@hunterhu3000 please rebase to pick up CI fixes

File: drivers/counter/counter_mchp_xec.c (hunk 4 above)
Comment (albertofloyd) [reply drivers/counter/counter_mchp_xec.c]: It's SoC specific maybe the change could be changed to CONFIG_SOC_SERIES_MEC172X

Comment (hunterhu3000) [reply drivers/counter/counter_mchp_xec.c]: This is changed in the latest version

**Review by MaureenHelm — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**